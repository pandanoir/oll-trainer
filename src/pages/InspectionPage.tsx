import { useCallback, useEffect, useState } from 'react';
import { useHotkeys } from 'react-hotkeys-hook';
import 'twin.macro';

import eightSecondsSoundUrl from '../sound/eightSeconds.mp3';
import twelveSecondsSoundUrl from '../sound/twelveSeconds.mp3';
import timeoutSoundUrl from '../sound/警告音1.mp3';
import dnfSoundUrl from '../sound/警告音2.mp3';
import { playAudio, playSilence } from '../utils/playAudio';

const showRestInspectionTime = (pastMillisec: number) =>
  15 * 1000 >= pastMillisec
    ? `${Math.ceil(15 - pastMillisec / 1000)} seconds`
    : 17 * 1000 >= pastMillisec
    ? '+2'
    : 'DNF';

const fetchAudio = (url: string) =>
  fetch(url).then((response) => response.arrayBuffer());
const eightSecondsSound = fetchAudio(eightSecondsSoundUrl);
const twelveSecondsSound = fetchAudio(twelveSecondsSoundUrl);
const timeoutSound = fetchAudio(timeoutSoundUrl);
const dnfSound = fetchAudio(dnfSoundUrl);

type StopwatchMode = 'in inspection' | 'wait for user input';

const usePlayOnce = (audio: Promise<ArrayBuffer>) => {
  const [hasPlayed, setHasPlayed] = useState(false);

  return {
    play: useCallback(async () => {
      if (!hasPlayed) {
        setHasPlayed(true);
        playAudio(await audio);
      }
    }, [audio, hasPlayed]),
    reset: useCallback(() => {
      setHasPlayed(false);
    }, []),
  } as const;
};
export const InspectionPage = () => {
  const [pastMillisec, setPastMillisec] = useState(0);
  const [mode, setMode] = useState<StopwatchMode>('wait for user input');
  const { play: call8Seconds, reset: resetHasCalled8Seconds } =
    usePlayOnce(eightSecondsSound);
  const { play: call12Seconds, reset: resetHasCalled12Seconds } =
    usePlayOnce(twelveSecondsSound);
  const { play: callTimeout, reset: resetHasCalledTimeout } =
    usePlayOnce(timeoutSound);
  const { play: callDNF, reset: resetHasCalledDNF } = usePlayOnce(dnfSound);

  useHotkeys(
    'space',
    () => {
      playSilence();
      if (mode === 'wait for user input') {
        setMode('in inspection');
      } else if (mode === 'in inspection') {
        setMode('wait for user input');
      }
    },
    [mode]
  );
  useEffect(() => {
    if (mode === 'in inspection') {
      const startAt = performance.now();
      resetHasCalled8Seconds();
      resetHasCalled12Seconds();
      resetHasCalledTimeout();
      resetHasCalledDNF();
      const id = setInterval(() => {
        setPastMillisec(performance.now() - startAt);
      }, 1000 / 30);
      return () => {
        clearInterval(id);
        setPastMillisec(0);
      };
    }
  }, [
    mode,
    resetHasCalled8Seconds,
    resetHasCalled12Seconds,
    resetHasCalledTimeout,
    resetHasCalledDNF,
  ]);
  useEffect(() => {
    if (pastMillisec < 8e3) {
      return;
    }
    if (pastMillisec < 12e3) {
      call8Seconds();
    } else if (pastMillisec < 15e3) {
      call12Seconds();
    } else if (pastMillisec < 17e3) {
      callTimeout();
    } else {
      callDNF();
    }
  }, [call8Seconds, call12Seconds, callTimeout, callDNF, pastMillisec]);

  return (
    <div tw="flex flex-col flex-1 space-y-3">
      <h1 tw="text-2xl">inspection</h1>
      <div
        tw="flex flex-col justify-center items-center flex-1"
        onClick={() => {
          playSilence();
          if (mode === 'wait for user input') {
            setMode('in inspection');
          } else if (mode === 'in inspection') {
            setMode('wait for user input');
          }
        }}
      >
        {mode === 'wait for user input' ? (
          <span tw="select-none text-blue-400 font-bold">
            press space or tap to start inspection.
          </span>
        ) : (
          <div tw="flex flex-col space-y-1 select-none text-center">
            <span tw="text-5xl font-bold">
              {showRestInspectionTime(pastMillisec)}
            </span>
            <span tw="text-gray-700 dark:text-gray-400">
              esc or tap to reset inspection.
            </span>
          </div>
        )}
      </div>
    </div>
  );
};
