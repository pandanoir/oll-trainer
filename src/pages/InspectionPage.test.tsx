/**
 * @jest-environment jsdom
 */
import { render, act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { InspectionPage } from './InspectionPage';

jest.mock('../sound/steady.mp3', () => '');
jest.mock('../sound/eightSeconds.mp3', () => '');
jest.mock('../sound/twelveSeconds.mp3', () => '');
jest.mock('../sound/警告音1.mp3', () => '');
jest.mock('../sound/警告音2.mp3', () => '');

jest.mock('../utils/playAudio', () => ({
  playAudio: jest.fn(),
  playSilence: jest.fn(),
}));

jest.setTimeout(50000);
describe('InspectionPage', () => {
  it('snapshot', () => {
    const { asFragment, queryByText, getByText } = render(
      <BrowserRouter>
        <InspectionPage />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(queryByText(/15 seconds/)).toBe(null);
    act(() => {
      getByText(/press space or tap to start inspection/).click();
    });
    expect(asFragment()).toMatchSnapshot();
    expect(queryByText(/15 seconds/)).not.toBe(null);
    act(() => {
      jest.advanceTimersByTime(1000 + (1000 / 60) * 2);
    });
    expect(queryByText(/14 seconds/)).not.toBe(null);

    act(() => {
      jest.advanceTimersByTime(14000);
    });
    expect(queryByText(/\+2/)).not.toBe(null);
  });
});
