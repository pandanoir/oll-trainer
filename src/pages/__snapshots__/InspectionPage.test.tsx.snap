// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspectionPage snapshot 1`] = `
<DocumentFragment>
  <div
    class="_StyledDiv-sc-1hem0f1-0 kkHgeD"
  >
    <h1
      class="_StyledH-sc-1hem0f1-1 dUMIYo"
    >
      inspection
    </h1>
    <div
      class="_StyledDiv2-sc-1hem0f1-2 TXdDb"
    >
      <span
        class="_StyledSpan-sc-1hem0f1-3 jxPfvm"
      >
        press space or tap to start inspection.
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InspectionPage snapshot 2`] = `
<DocumentFragment>
  <div
    class="_StyledDiv-sc-1hem0f1-0 kkHgeD"
  >
    <h1
      class="_StyledH-sc-1hem0f1-1 dUMIYo"
    >
      inspection
    </h1>
    <div
      class="_StyledDiv2-sc-1hem0f1-2 TXdDb"
    >
      <div
        class="_StyledDiv3-sc-1hem0f1-4 fNrAjv"
      >
        <span
          class="_StyledSpan2-sc-1hem0f1-5 bKnFds"
        >
          15 seconds
        </span>
        <span
          class="_StyledSpan3-sc-1hem0f1-6 bFyvHq"
        >
          esc or tap to reset inspection.
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
